### ISSUE TITLE
Low overhead for `kprove` (call `kompile` ahead of time)

### ISSUE THREAD

**Issue by denis-bogdanas** [runtimeverification/k#1333]
File `demo.k`
```
module DEMO
  imports DOMAINS
  configuration <k> .K </k>
endmodule

module EXTENDED
  imports DEMO
  configuration <extended> <k/> </extended>
endmodule
```

File `extended-bad-module-spec.k`
```
requires "demo.k"

module VERIFICATION-DEMO
    imports DEMO
endmodule

module VERIFICATION-EXTENDED
    imports EXTENDED
    imports VERIFICATION-DEMO
endmodule

module EXTENDED-BAD-MODULE-SPEC
    imports VERIFICATION-EXTENDED

    rule <k> .K => ?_ </k>
endmodule
```
File `Makefile`
```
DEF=extended
EXT=test
TESTDIR=.
KOMPILE_BACKEND=haskell
KPROVE_FLAGS=--def-module VERIFICATION-DEMO

export KOMPILE_BACKEND
export KPROVE_FLAGS

include ../../../include/kframework/ktest.mak
```

**Comment by ehildenb** [comment]
@denis-bogdanas we need more information on how to reproduce. Is this a `kompile` time error? And what is the error message? What do you expect to happen.

I see you opened the PR with the test, but please just post that info here.

**Comment by denis-bogdanas** [comment]
We have a spec that imports `VERIFICATION-EXTENDED`, which imports `VERIFICATION-DEMO`. And kprove argument is `--def-module VERIFICATION-DEMO` instead of `--def-module VERIFICATION-EXTENDED`. The 2 verification modules have different configuration.

I expect an error telling that `--def-module` argument is incorrect, because the definition top cell is different from the spec rule top cell.

Current output:
```
java.lang.AssertionError: Unexpected non-rule claim syntax KItem ::= Cell
     [exec] 	at org.kframework.backend.kore.ModuleToKORE.convertSpecificationModule(ModuleToKORE.java:845)
```

This is low priority.

**Comment by dwightguth** [comment]
the entire way definitions are constructed in the kprove frontend is nasty, and it's the reason for this type of problem. We need to have a serious discussion about how to change the way the kprove and kompile interact, which will probably introduce breaking changes to how specs are written, but I don't see a better option.

**Comment by denis-bogdanas** [comment]
I agree. Some time ago I proposed replacing `--def-module` option with with a module attribute. Can't find that discussion. But I see this topic is popping once in a while: https://runtimeverification.slack.com/archives/C7E701MFG/p1567285075096300

It eats a slice of our productivity, particularly when things get incorrectly configured in make scripts and we are scratching our head at never seen errors.

If changes to specs are needed, they will be mechanical, I don't think it's an issue.

**Comment by ehildenb** [comment]
Ok, then maybe we can use this as a chance to propose some concrete changes to how KProve builds definitions for proving.

The only thought I've had on this is that the module with proofs should always import _exactly_ one module, and that should be the proving module. But I'm not sure if that is the type of overhaul that you guys are thinking about.

Another idea I've had is to get rid of the proving module altogether, and just make a disntinction between `rule` and `claim`. Then you can mix in `claim` with any module, and it will be proved in the module that you mix it in with, then can be added as a lemma for other modules which import that one. I think this gives the most modularity and guarantee, but is a pretty big change.

@denis-bogdanas what are you thoughts? @dwightguth? And maybe @daejunpark as well? Maybe there is some small change we can make to improve the situation.

**Comment by denis-bogdanas** [comment]
@grosu argued that we might want to have 2 modules with an equivalent set of rules. And prove one module relative to the other one. Then the other way around. Also consider that we might want to have claims distributed into several modules.

Having the definition module marked with an attribute seems like a clean solution that fits all scenarios above and doesn't restrict file structure.

Then @grosu scenario would look like:
```
module MOD1
  import DEF1
  rule rule1
endmodule

module DEF1 [definition]
  import MOD2
endmodule

module MOD2
  import DEF2
  rule rule2
endmodule

module DEF2 [definition]
  import MOD1
endmodule
```
```
kprove -m MOD1 file.k
kprove -m MOD2 file.k
```

**Comment by ehildenb** [comment]
That doesn't work in all cases. What if you import multiple modules marked as `definition`? I would much rather have the convention where we say the proving module is only allowed to import a single module (the main definition module).

Another thing to chew on, we often times do proving with the same `VERIFICATION` module, and each time we invoke `kprove`, it will have to re-kompile that module. Could we instead ahead-of-time kompile it and use it for doing the proving @dwightguth ?

**Comment by ehildenb** [comment]
I'm changing this issue to better reflect that it's now a discussion about KProve architecture.

**Comment by denis-bogdanas** [comment]
We can check that only one of directly imported modules has `[definition]`.

**Comment by ehildenb** [comment]
No, I'm saying that you may actually _want_ to have two different modules with `[definition]` on them, so I think your solution impedes that (strictly less functionality).

```
module A [definition] endmodule

module B [definition] endmodule

module C imports A imports B endmodule

module D [definition] endmodule
```

Now we cannot do verification using module `D` as the main module, though it may very well be reasonable to do so (eg. if you're combining two separate definitions and want to do verification in the joint definition).

**Comment by denis-bogdanas** [comment]
If there is actual need to combine several definition modules, we can either support multiple modules with `[definition]`, or have one auxiliary module that imports definition modules:
```
module A imports DEF endmodule //spec
module DEF-AUX [definition] imports DEF1 imports DEF2 endmodule
```

We discuss 2 different things: (1) how to pass to kprove the information what module(s) represent definition, and (2) how to compile the definition. My proposal is for point 1: to replace current `--def-module` with `[definition]`. So far it can adapt point 2 needs.

**Comment by dwightguth** [comment]
What I would really like to do is remove kompile functionality from kprove entirely, make it a requirement that kprove only take specification modules as input, then entirely remove vdefinition.kore from the pipeline.

What this is going to require though is finding a way to modularly store the cached parses of rules to be reused across compiling separate definitions that use an overlapping set of modules, and then have the person doing proofs first call kompile, and then kprove if they want to add custom lemmas and functions and such.

There are serious problems with the current approach where we patch together a definition from different sources after the fact, as evidenced by this type of bug.

With that being said, if we do what I describe, the overhead of invoking the same proof several times, or of modifying the specification and invoking the proof again, or of using the same definition for a bunch of different proofs, will go down substantially. We will also at the same time get a much cleaner implementation without the current garbage fire which is ProofDefinitionBuilder.

**Comment by ehildenb** [comment]
I agree with calling `kompile` ahead of time for proving, that seems like a sane solution to me. If we have that, then all we need to do is `kprove --definition path/to/directory/with/*-kompiled`, and won't need to supply `--def-module` anymore right? Because `--def-module` will be just the main module of the definition?

**Comment by dwightguth** [comment]
Yes, the idea is that instead of creating any kind of vdefinition.kore, you would just pass -d some-definition and it would use exactly what was already compiled. Basically this is the outgrowth of my realization that the design direction we went in in the past for the kprove frontend was headed in the wrong direction and we need to reverse course somewhat. That being said, I understand why we made that decision in the first place, and I do want to make sure we can create an experience that is at least as good as we have now. That's why I want to try to come up with a way for compiling some additional modules and then using them in the proof to be as modular and fast as possible. But essentially, my take is that having kprove bake in a bunch of parsing and compilation it wasn't really meant to be doing has caused us nothing but headaches.

**Comment by ehildenb** [comment]
Ok, I think we can have a discussion about this at the next K frontend meeting. @denis-bogdanas and @daejunpark may both be interested in attending.

**Comment by denis-bogdanas** [comment]
@dwightguth What you propose, without parse caches, seems not so hard to achieve. From K user perspective we would have several versions of the same spec: one for concrete execution and one or more for proofs. Just like there are several versions of KEVM right now. Then lemmas would simply be added to the definition end and compiled with kprove. `kprove` would only have to parse the modules with proofs. But this doesn't eliminate `ProofDefinitionBuilder` entirely as you still have to parse proofs, and kprove doesn't have access to them.

So you can think of implementing the right interface first, then caching.

**Comment by daejunpark** [comment]
I like the idea of clearly separating kompile from kprove, especially if it can significantly reduce the frontend time of kprove.  (Currently, the frontend time of kprove takes 2-3 minutes for KEVM specs, and I couldn't be happier if it could be reduced down to seconds!)

So, if I understood correctly, the usage will be: we kompile a definition module in advance, and then kprove will just take the `--main-module` module passed to kompile as the definition module, and will take any rules outside the definition module as claims, right?

**Question 1**. In that case, kprove will parse-fail if it finds any syntax declarations outside the definition module, right?  What if it finds lemma rules (i.e., rules over [function] symbols) outside the definition module?  Will it parse-fail, or consider them as claims and try to prove (and probably would fail)?  The former is better, but I'm not sure how hard to implement it.

----

Now, I think the only remaining question is for the case when we have different specs that require different set of lemmas.

A straightforward option is to kompile a baseline definition multiple times with different set of lemmas. For example, if we have:

def.k:
```
module BASE-WITH-LEMMA-A
  imports BASE
  imports LEMMA-A
endmodule

module BASE-WITH-LEMMA-B
  imports BASE
  imports LEMMA-B
endmodule

module BASE-WITH-LEMMA-A-B
  imports BASE
  imports LEMMA-A
  imports LEMMA-B
endmodule

module BASE ... endmodule

module LEMMA-A ... endmodule
module LEMMA-B ... endmodule
```

- a-spec.k: `module A-SPEC   imports BASE-WITH-LEMMA-A   ... endmodule`
- b-spec.k: `module B-SPEC   imports BASE-WITH-LEMMA-B   ... endmodule`
- c-spec.k: `module C-SPEC   imports BASE-WITH-LEMMA-A-B   ... endmodule`

Then, we can do like:
```
$ kompile def.k --main-module BASE-WITH-LEMMA-A -d a
$ kprove a-spec.k -d a

$ kompile def.k --main-module BASE-WITH-LEMMA-B -d b
$ kprove b-spec.k -d b

$ kompile def.k --main-module BASE-WITH-LEMMA-A-B -d c
$ kprove c-spec.k -d c
```

**Question 2.**  Is it possible to allow the following by any chance?
```
$ kompile def.k --main-module BASE-WITH-LEMMA-A-B

$ kprove a-spec.k --def-module BASE-WITH-LEMMA-A
$ kprove b-spec.k --def-module BASE-WITH-LEMMA-B
$ kprove c-spec.k --def-module BASE-WITH-LEMMA-A-B
```
It allows us to run kompile once, but run kprove with a different definition (sub-)module.  This will be still different from the current behavior of kprove, because it will not allow a spec module to introduce any new syntax declarations, or new lemma rules, but just allow kprove to use any sub modules already kompiled.

**Comment by ehildenb** [comment]
So it seems that everyone is on-board with removing `kompile` functionality from `kprove`, which adds the hard restriction that you _cannot_ add any new syntax declarations to the module that is to be proven.

But what we are not sure about is should we still allow a new module with lemmas to be added to `kprove`? My gut says "yes", because then we can reduce how many definitions we need to kompile ahead of time and as long as those new modules with lemmas don't introduce new syntax, parsing them should be very similar to just parsing the module which has the claims itself.

**Comment by dwightguth** [comment]
Some input is needed from the haskell backend team, I think. I see no reason we wouldn't be able to parse both claims and lemmas in the specification modules, as long as no new syntax is added. It ought even to be possible to add new productions as long as they have the `token` attribute and don't create a new sort. But they would go through a different set of compilation steps because they wouldn't be part of the "definition". Whether those different set of compilation steps still produces an output that can be plugged into kprove along with the original definition, is less clear to me. It depends somewhat on the capabilities of the haskell backend. What forms do lemmas take in the haskell backend? would it be possible to use the definition produced by `kompile` and then simply add in the lemmas as a separate file? why or why not? @ttuegel needs to answer I think.

**Comment by ttuegel** [comment]
> What forms do lemmas take in the haskell backend? would it be possible to use the definition produced by `kompile` and then simply add in the lemmas as a separate file? why or why not?

I'm not 100% sure I understand the question, but I'll try to answer. Lemmas for functions take the same form as simplification rules included in the definition. It should be possible to add these in a separate file, unless they would change the attributes attached to some symbol. (I think only `anywhere` rules change the symbol attributes, though.) There are also trusted claims, which are "like" lemmas, but they just go through the usual claim compilation pipeline; I don't think anything proposed here would change that.

Basically, as long as none of the additions change attributes of previously-declared sorts or symbols, it should be alright to chop the definition and spec into as many files as we want. It will require some interface-level changes to the backend, but nothing else.

**Comment by ehildenb** [comment]
Because of how the Haskell backend currently works, we suspect that it should be OK to have intermediate modules between `MAIN-MODULE` and the verification module as long as those modules _only_:

-   Declare new equation-type rules which specifically have the `simplification` attribute (specifically, no `priority` or `owise` attributes, recall that you can now give the `simplification(_)` attribute an integer giving priority to `simplification` rules too).
-   Declare new claims (we may want claims in submodules for organizing lemmas of top-level proof, proving one lemma at a time, and then have the top-level module re-use the lemmas in submodules).
-   Declare productions in existing sorts with the `token` attribute.

**Comment by ttuegel** [comment]
> Because of how the Haskell backend currently works

To clarify, this is not due to the backend, but how the frontend works. Adding other types of syntax or rules in new modules can change definitions that were already closed when another module was compiled.

**Comment by ehildenb** [comment]
A simpler path (initially) may be:

-   Modify build systems of semantics to allow building the VERIFICATION module directly (`kompile .... tests/specs/mcd/verification.k`)
-   Modify build system of semantics to call `kompile ... tests/specs/mcd/verification.k` before calling `kprove`.
-   Add an option to `kprove` called `--skip-kompile`, which just looks for the correct kompiled directory and skips compiling the def-module and just loads it, then continues with parsing the spec rules.

**Comment by ehildenb** [comment]
@dwightguth points out that `--skip-kompile` would largely be the same amount of work after the `claim` keyword is introduced anyway, so we might as well go forward with the full solution.
